Stop considering that Negative Binomial has a dispersion parameter

nalimilan opened the pull request:
This is just incorrect. Fix tests to reflect correct degrees of freedom and therefore AIC, BIC and standard errors.

This requires adapting how `negbin` works as we need to count one additional degree of freedom due to the shape parameter (distinct from the dispersion, which is 1) being estimated from the data. Do this by wrapping the shape parameter in an `Estimated` type that we use to detect that degrees of freedom need adjusting.

Also avoid defining `dispersion_parameter` fallback for any distribution: it's clearer for users to get a `MethodError` so that they have to define the appropriate method themselves, than silently returning a wrong value.

Fixes #564, #617.


--------

I have tentatively implemented the first of three possible approaches I see to the `negbin` issue to illustrate how it looks:
1. Wrap the shape parameter of `NegativeBinomial` in an `Estimated` type: I find this is somewhat elegant. The disadvantage is that users can't really work with that parameter if they extract it. We could define more methods on this type but that would be weird anyway. We'd better provide an accessor which would unwrap it, which is needed anyway.
2. Add a field to `GeneralizedLinearModel` (or one of its component) indicating that one parameter was estimated. This has the advantage of being super simple. The drawback is that its a bit ad-hoc but in practice it probably doesn't matter.
3. Make `negbin` return a `NegativeBinomialModel <: GLMLikeModel <: LinPredModel`. This has the advantage of clarity (in R also `glm.nb` returns an object of a different class) since these are not really GLMs. In terms of implementation I'm not sure whether it would be easy (just replace `AbstractGLM` with `GLMLikeModel` for most methods) or not.

Opinions on the best approach?

Note that R's glm seems broken with `negative.binomial` as [it estimates a dispersion parameter](https://stats.stackexchange.com/questions/70619/dispersion-in-summary-glm) even if it's 1 by theory. You have to do `summary(m, dispersion=1)` to get the correct standard errors. Yet `?negative.binomial` claims it's meant do be used with `glm` with a fixed shape.

nalimilan committed:
Stop considering that Negative Binomial has a dispersion parameter

This is incorrect when the shape parameter has been fixed manually and the
model estimated via `glm`. Fix tests to reflect correct degrees of freedom
and therefore AIC, BIC and standard errors.

This requires adapting how `negbin` works as we need to count one additional
degree of freedom due to the shape parameter (distinct from the dispersion,
which is 1) being estimated from the data. Do this by returning a
`NegativeBinomialModel` instead of a standard GLM.

andreasnoack commented:
The two cases here are different, and I think it is best to handle them in separate PRs. `Geometric` only has a single parameter while `NegativeBinomial` has two parameters where the second isn't a dispersion parameter in the GLM sense. I agree that we should not have a fallback but only define `dispersion_parameter` for the cases we want to support. I think the better way to handle the NegativeBinomial case is to introduce a new type just for NegativeBinomial, somewhat like what MASS does, and I think it should just be a direct subtype of `LinPredModel`, i.e. no need for a new abstract type.

nalimilan committed:
Stop considering that Geometric has a dispersion parameter

This is just incorrect. Fix tests to reflect correct degrees of freedom
and therefore AIC, BIC and standard errors.

Also avoid defining `dispersion_parameter` fallback for any distribution:
it's clearer for users to get a `MethodError` so that they have to define
the appropriate method themselves, than silently returning a wrong value.

`NegativeBinomial` will be fixed separately as it requires a deeper
refactoring of `negbin` (#624).

nalimilan commented:
It's hard to fix the two cases completely separately as we have tests checking that geometric and negative binomial give the same results, but we can leave some tests as broken temporarily. Let's start by fixing `Geometric`, which is the simplest: #633.

nalimilan committed:
Fix CI

nalimilan committed:
Try triggering GitHub Actions

nalimilan committed:
Merge branch 'master' into nl/dispersion

nalimilan committed:
Fix merge

andreasnoack commented on src/glmfit.jl:
```diff
@@ -668,14 +676,20 @@ function glm(formula::FormulaTerm, data, d::UnivariateDistribution,
 generally written σ for linear models and ϕ for generalized linear models.
-It is, by definition, equal to 1 for the Bernoulli, Binomial, and Poisson families.
+
+It is, by definition, equal to 1 for the Bernoulli, Binomial, Geometric, Negative
```
I don't think it is correct to say that it is one for the negative binomial. As I understand McCullagh and Nelder, the proper dispersion parameter is only defined in the case when the variance can be written
```math
\mathrm{Var}(Y) = \frac{\phi}{m}\mathrm{Var}(\mu)
```
in which case $\phi$ is called the dispersion parameter. For the negative binomial, the variance is
```math
\mathrm{Var}(Y) = \mu + \mu^2/k
```
so $k$ isn't a *dispersion* parameter according to their definition, but I wouldn't say it is one either. It happens to be the case the Fisher information is $\phi(X^T W X)^{-1}$ for the proper GLMs and for negative binomial, it is just $(X^T W X)^{-1}$ so you can reuse the code for the GLMs if you pass $\phi=1$ but that is just a computational convenience and not because the negative binomial fits into the framework.

nalimilan replied:
So say "Negative Binomial (with fixed θ)"?

andreasnoack commented on src/lm.jl:
```diff
@@ -208,6 +208,13 @@ function lm(f::FormulaTerm, data;
 end
 
+"""
+    dof(model::GeneralizedLinearModel)
```
```suggestion
    dof(model::LinearModel)
```

andreasnoack commented on src/glmfit.jl:
```diff
@@ -394,7 +394,15 @@ function nullloglikelihood(m::GeneralizedLinearModel)
+For generalized linear models, consumed degrees of freedom correspond
+to the number of estimated coefficients, plus one for the estimated
+dispersion parameter if the distribution is `Gamma`, `InverseGaussian` or `Normal`
+(see [GLM.dispersion_parameter](@ref)).
```
Are you sure it is useful to loosen the signature from GeneralizedLinearModel to AbstractGLM? Isn't the only type that is an `AbstractGLM` but not a `GeneralizedLinearModel` the new `NegativeBinomialModel` for which there is a separate model?

nalimilan commented on src/negbinfit.jl:
```diff
@@ -1,3 +1,25 @@
+mutable struct NegativeBinomialModel{G<:GlmResp,L<:LinPred} <: AbstractGLM
```
With a `getproperty` override to delegate field accesses? Doesn't really seem simpler in the end?
